### ISSUE TITLE
Invoice links missing from CSV

### ISSUE THREAD

**Issue by mgiuffrida** [philipmulcahy/azad#121]
When "Show invoice links" is checked, downloading a CSV from the result doesn't include invoice links.

The `invoice` column is present at the end of the header line, but that field is always empty (all of the rows end with a comma). Not sure if this is a known bug; is the `invoice` column a placeholder for forward compatibility?

**Comment by philipmulcahy** [comment]
Did you set the checkbox before clicking on the year button?
If you’ve already got the year’s orders shown in the table, the csv button is going to take its info from that table, even if you flip the checkbox before.
Maybe the checkbox label should say “fetch” instead of “show”...

**Comment by mgiuffrida** [comment]
Yes, the checkbox was checked before starting.

**Comment by philipmulcahy** [comment]
…and you could see invoice links in the on-screen table?

**Comment by mgiuffrida** [comment]
The `payments` field has links in the table ([example](https://smile.amazon.com/gp/css/summary/print.html/ref=oh_aui_ajax_pi?ie=UTF8&orderID=111-0205632-8073066)) to pages that look like receipts.

The `invoice` field is empty in the table (and in the csv).

Note that the payment links aren't included in the csv either, just the text of that field.

**Comment by philipmulcahy** [comment]
sorry for being slow on this. Would you mind attaching a console log here?

**Comment by mgiuffrida** [comment]
Chrome is being weird and not letting me "save as" in the console, so this was copied and pasted:

[console-log.txt](https://github.com/philipmulcahy/azad/files/5165229/console-log.txt)

**Comment by philipmulcahy** [comment]
I just looked at my .co.uk 2010, and it fails to fetch any invoices.
The first order I drilled into has this:
![image](https://user-images.githubusercontent.com/1668892/92159318-1dd63680-ee25-11ea-9a58-4ab02f6bf103.png)

The second:
![image](https://user-images.githubusercontent.com/1668892/92159582-8c1af900-ee25-11ea-9bab-0fd28cd90609.png)

I followed the second order's link and it contained:
![image](https://user-images.githubusercontent.com/1668892/92159681-b076d580-ee25-11ea-8e63-a8094d0f0a68.png)
Could your .com 2010 order set all be in the same situations (e.g. unavailable or requiring an async request possibly involving a human)?

Yours,

Philip

**Comment by mgiuffrida** [comment]
this happens for every .com year, 2010 just had a shorter order list. I can access my invoices normally.

**Comment by philipmulcahy** [comment]
sorry - I got confused between this problem and https://github.com/philipmulcahy/azad/issues/124
I'm wondering if my next move is adding some more logging...

**Comment by philipmulcahy** [comment]
Hi mgiuffrida@,

Would you mind doing a quick warm up experiment?

1) Download http://www.mulcahyfamily.org/azad/azad.986a0386c.zip
2) unzip it
3) open chrome://extensions
4) disable the store version of azad
5) turn developer mode on
6) Load Unpacked Extension
7) point at the unzipped folder
8) open an amazon tab 
9) do what you need to do to get a csv
10) look for a line in the log like this: value_done_promises.length 1353

Repeat 9-10 and see if the number changes (not expected)
Is the number a multiple of 11 (the number of columns there should be if the invoice links feature is enabled) ?
I was wondering if the csv generation isn't waiting for all of the promises  that draw the table cells to resolve before copying them out (e.g. a race), but it seems like a bit of a long shot given the behaviour I'm seeing on my machine.

Thanks,

Philip

**Comment by philipmulcahy** [comment]
ping

**Comment by philipmulcahy** [comment]
no response from user - closing

**Comment by ksbrar** [comment]
Also having this issue. Can I help resolve? @philipmulcahy

**Comment by philipmulcahy** [comment]
@ksbrar - can you please
1)open a new ticket
2), include a console log (as an attachment - drag it to a new post on the ticket) for an entire session where you reproduce the problem (e.g from when you open the amazon page to when the bad thing happens or the good thing doesn't happen
3) email an order debug json file to azadextension@gmail.com with the ticket link somewhere in the email - choose an order that doesn't have an invoice link you think should be there, right click on the order link in the output table and select the debug file option (it has the azad logo next to it. It should appear as a download after a few seconds.
Yours,
Philip
